core/vm/logger.go

```diff
@@ -142,27 +141,32 @@ func (l *StructLogger) CaptureStart(from common.Address, to common.Address, crea
 
 // CaptureState logs a new structured log message and pushes it out to the environment
 //
-// CaptureState also tracks SSTORE ops to track dirty values.
+// CaptureState also tracks SLOAD/SSTORE ops to track storage change.
 func (l *StructLogger) CaptureState(env *EVM, pc uint64, op OpCode, gas, cost uint64, memory *Memory, stack *Stack, rStack *ReturnStack, contract *Contract, depth int, err error) error {
 	// check if already accumulated the specified number of logs
 	if l.cfg.Limit != 0 && l.cfg.Limit <= len(l.logs) {
 		return errTraceLimitReached
 	}
-
 	// initialise new changed values storage container for this contract
 	// if not present.
-	if l.changedValues[contract.Address()] == nil {
-		l.changedValues[contract.Address()] = make(Storage)
+	if l.storage[contract.Address()] == nil {
+		l.storage[contract.Address()] = make(Storage)
 	}
-
-	// capture SSTORE opcodes and determine the changed value and store
-	// it in the local storage container.
+	// capture SLOAD opcodes and record the read entry in the local storage
+	if op == SLOAD && stack.len() >= 1 {
+		var (
+			address = common.Hash(stack.data[stack.len()-1].Bytes32())
+			value   = env.StateDB.GetState(contract.Address(), address)
+		)
+		l.storage[contract.Address()][address] = value
+	}
+	// capture SSTORE opcodes and record the written entry in the local storage.
 	if op == SSTORE && stack.len() >= 2 {
 		var (
 			value   = common.Hash(stack.data[stack.len()-2].Bytes32())
 			address = common.Hash(stack.data[stack.len()-1].Bytes32())
 		)
-		l.changedValues[contract.Address()][address] = value
+		l.storage[contract.Address()][address] = value
 	}
```

Comment:
Any reason not to put this entire clause behind a `if !l.cfg.DisableStorage` ?

Reply:
right, fixed